### PR TITLE
Adds npm for managing client-side dependencies

### PR DESCRIPTION
**Pull request by flashbackzoo**
This is an implementation of https://github.com/silverstripe/silverstripe-framework/issues/4372

Not all libraries are handled by npm yet. We're still on really old versions of libraries aren't available on npm and/or there is no `package.json` file on GitHub. More recent versions _are_ available on npm / GitHub and we've got stories in our internal backlog to upgrade these old libraries.

So this PR includes...

- A package.json file with build and CI scripts
- Shrinkwrap file
- Gulp for running build tasks
- A 'build' task for copying library files from node_modules to thirdparty
- A 'sanity' task for makes sure library files in thirdparty match what's in node_modules
- Updates to .travis.yml for running the sanity task as part of CI

### CODE REVIEW

**Comment by sminnee** [comment]
If node_modules is ignored, how can users of SilverStripe use it without installing NPM? Only people developing on SilverStripe should need to install NPM.

I believe that the solution to this is to check node_modules into the repo, and have some kind of post-commit hooks for this.

File: gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// Make sure the version of Node being used is valid.
+if (!semver.satisfies(process.versions.node, packageJson.engines.node)) {
+    console.error('Invalid Node.js version. You need to be using ' + packageJson.engines.node);
+    process.exit(1);
+}
+
+gulp.task('build', function () {
+    copyFiles(blueimpFileUploadConfig);
```
Comment (sminnee) [inline gulpfile.js]: Will we need extra lines in here for every library? That seems suboptimal.

Comment (flashbackzoo) [inline gulpfile.js]: Yeah we'll need one line per library. I did consider this and think it's the best solution, given copying files to thirdparty will become obsolete once we've upgraded our library versions, and https://github.com/silverstripe/silverstripe-framework/issues/4606 is implemented.

The other option was to have an object (or even better a separate json file) which has file / path mappings for all libraries, and everything in there gets copied to thirdparty automagically. I felt like having an extra config file was over engineering seeing it's a stopgap solution that will be super-seeded fairly soon. But I'm happy to reconsider if others feel the same way...

Comment (sminnee) [inline gulpfile.js]: Ok, if it's just a short-term fix, that's okay. Bring on #4606!

**Comment by sminnee** [comment]
Given the implications of this change, I'd like to see 3 core committers give this a :+1: before we merge.

:+1: from me.

**Comment by tractorcow** [comment]
Why can't I see travis results for this PR? Is the .travis.yml broken?

**Comment by tractorcow** [comment]
Yep, it's borked. :dancer: 
image

File: .travis.yml [hunk 2]
```diff
@@ -8,6 +8,7 @@ addons:
       - tidy
 
 env:
+  - TRAVIS_NODE_VERSION="4.2.0"
```
Comment (tractorcow) [inline .travis.yml]: Invalid yml here; That needs to go into global:

File: .travis.yml [hunk 3]
```diff
@@ -41,6 +44,7 @@ before_script:
  - php ~/travis-support/travis_setup_php54_webserver.php --if-env BEHAT_TEST
 
 script:
+ - npm run sanity
```
Comment (tractorcow) [inline .travis.yml]: Do we need this every run? Why not just on, for example, BEHAT_TEST = 1 ?

File: .travis.yml (hunk 2 above)
Comment (flashbackzoo) [inline .travis.yml]: Ahh thanks. I'll fix that up.

File: .travis.yml (hunk 3 above)
Comment (flashbackzoo) [inline .travis.yml]: Hmmm it's not really related to Behat but I see what you mean. Maybe a new variable `JAVASCRIPT_TESTS` (or something similar) which runs the sanity check, and when we have them, JS unit tests?

**Comment by flashbackzoo** [comment]
@tractorcow I've fixed `.travis.yml` and added a `JAVASCRIPT_TEST` flag to run the sanity check. This can also be used for JS unit tests when we have them :smile:

File: package.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "silverstripe-framework",
+  "version": "4.0.0",
+  "description": "The SilverStripe framework",
+  "directories": {
+    "doc": "docs",
+    "test": "tests"
+  },
+  "engines": {
+    "node": "~4.2.0"
```
Comment (chillu) [inline package.json]: Assuming node follows semver, isn't this too strict? We should require `~4`, so anything lower than 5.x.x.

File: docs/en/05_Contributing/01_Code.md [hunk 5]
```diff
@@ -262,6 +262,38 @@ Helpful hint: You can always edit your last commit message by using:
 
 	$ git commit --amend
 
+## Working with client-side dependencies
+
+From time to time client-side dependencies like jQuery need to be upgraded, added, or removed. We have some tools for dealing with that.
```
Comment (chillu) [inline docs/en/05_Contributing/01_Code.md]: Make it clearer that this relates to core dependencies, and that users are free to use their own separate library versions in their own build systems.

File: package.json (hunk 4 above)
Comment (flashbackzoo) [inline package.json]: Node ~4.2.0 is LTS. That seemed like a good baseline...

File: docs/en/05_Contributing/01_Code.md (hunk 5 above)
Comment (flashbackzoo) [inline docs/en/05_Contributing/01_Code.md]: Cool - I'll update this.

**Comment by chillu** [comment]
Wow, that leaves a lot of JS deps to be upgraded, either because our release is too old and not tracked via NPM, or because we've "forked" the dep by making local modifications in `thirdparty/*`. Assuming that we won't completely rewrite the UI for 4.0, we'll still need important libraries like jQuery UI or Chosen.js though, so that work will have to be done at some point. But that doesn't block merging this PR, just flagging that the majority of work is still ahead of us. Here's a complete list from our internal backlog:

image

So a +1 from me.

**Comment by flashbackzoo** [comment]
@chillu I've updated the docs to make it clear the tooling only applies to core SilverStripe dependencies.

I also updated the error message you get when using an invalid Node.js version. It now includes 'If you want to manage multiple Node.js versions try https://github.com/creationix/nvm'

**Comment by chillu** [comment]
Sweet, three core committer votes (Sam, Damian, Ingo), merging!

**Pull request merged**